httpcore/_async/connection_pool.py

```diff
@@ -167,17 +182,50 @@ def _connection_acquiry_lock(self) -> AsyncLock:
     def _create_connection(
         self,
         origin: Tuple[bytes, bytes, int],
+        socket: AsyncSocketStream,
     ) -> AsyncHTTPConnection:
         return AsyncHTTPConnection(
             origin=origin,
-            http2=self._http2,
-            uds=self._uds,
+            socket=socket,
             ssl_context=self._ssl_context,
-            local_address=self._local_address,
-            retries=self._retries,
             backend=self._backend,
         )
 
+    async def _open_socket(
```

Comment:
Custom pool implementations would be able to override this method (along with a strict package pin) if they wanted to customize how sockets are opened.

Standalone implementations (single connection w/o pooling) could copy this code and use them in their own `open_custom_connection()` context manager…

Reply:
While you can always have your own implementation you also need to implement the underlying connection code which is not impossible but it would be nice to utilise the existing connection codes in httpcore. By reimplementing all the code you

* Duplicate work, it's IMO a non-trivial amount of code to get a basic HTTP connection going
* Missing out on new functionality/bugfixes that is added by httpcore who know what they are doing when it comes to issues
* Makes the implementers code easier to maintain, they don't need to worry about h11, h2 and how to interact with that

I totally understand this project is in it's infancy so you may not be prepared to expose the existing classes, I just wanted to point out something I wish to see in the future :)